### PR TITLE
Append tooltip to the closest positioned ancestor of the selection

### PR DESCRIPTION
**Pull request by Laurian**
Simple fix for #1, does not take into consideration nested positioned ancestors.

### CODE REVIEW

File: dist/sharect.babel.js [hunk 1]
```diff
@@ -27,6 +27,20 @@ var Sharect = function () {
     var _top = 0;
     var _left = 0;
 
+    function offsetParent(el) {
```
Comment (Laurian) [inline dist/sharect.babel.js:30]: This is basically https://api.jquery.com/offsetParent/ from https://github.com/jquery/jquery/blob/56136897f241db22560b58c3518578ca1453d5c7/src/offset.js#L162-L172

File: dist/sharect.babel.js [hunk 2]
```diff
@@ -79,6 +93,14 @@ var Sharect = function () {
       var DOCUMENT_SCROLL_TOP = window.pageXOffset || document.documentElement.scrollTop || document.body.scrollTop;
       _top = position.top + DOCUMENT_SCROLL_TOP - _iconSize - _arrowSize;
       _left = position.left + (position.width - _iconSize * _icons.length) / 2;
+      if (hasSelection()) {
```
Comment (Laurian) [inline dist/sharect.babel.js:96]: If selection's offsetParent is not `document.body`, adjust left and top positions with offsetParent positions

File: dist/sharect.babel.js [hunk 3]
```diff
@@ -103,16 +125,25 @@ var Sharect = function () {
 
       div.appendChild(arrow);
 
-      document.body.appendChild(div);
+      if (hasSelection()) {
```
Comment (Laurian) [inline dist/sharect.babel.js:128]: Appen tooltip to body or to nearest offsetParent() of selection; that makes the positioning simpler and with no need of adjusting it on scroll

File: dist/sharect.babel.js [hunk 4]
```diff
@@ -103,16 +125,25 @@ var Sharect = function () {
...
       function hasTooltipDrawn() {
-        return !!document.querySelector('.sharect');
+        var tooltip = document.querySelector('.sharect');
+        if (!tooltip) return false;
+
+        if (hasSelection() && offsetParent(tooltip) === offsetParent(_selection.anchorNode.parentElement)) {
+          return true;
+        }
+
+        tooltip.remove();
```
Comment (Laurian) [inline dist/sharect.babel.js:145]: If the selection is in a part of the page that has different offsetParent(), kill the tooltip and recreate.

File: src/sharect.js [hunk 5]
```diff
@@ -25,6 +25,20 @@ const Sharect = (function(){
     let _top = 0
     let _left = 0
 
+    function offsetParent(el) {
+      var offsetParent = el.offsetParent
```
Comment (estevanmaito) [inline src/sharect.js:29]: Use let instead of var

File: src/sharect.js [hunk 6]
```diff
@@ -40,8 +54,8 @@ const Sharect = (function(){
 
     function twitterButton() {
       const txt = _twitterConfig.username
-                  ? ` ${_twitterConfig.username} ${window.location.href}`
-                  : ` ${window.location.href}`
+                  ? ` ${_twitterConfig.username} ${encodeURIComponent(window.location.href)}`
```
Comment (estevanmaito) [inline src/sharect.js:57]: Don't do unrelated changes. It has nothing to do with the tooltip position.

File: src/sharect.js [hunk 7]
```diff
@@ -116,16 +138,28 @@ const Sharect = (function(){
 
       div.appendChild(arrow)
 
-      document.body.appendChild(div)
+      if (hasSelection()) {
+        offsetParent(_selection.anchorNode.parentElement).appendChild(div)
```
Comment (estevanmaito) [inline src/sharect.js:142]: This is the only line needed here.
`drawTooltip` is called after a `hasSelection` test in the `mouseup` event, so no need to test again here.
`offsetParent` already returns the `offsetParent || document.body`

File: src/sharect.js [hunk 8]
```diff
@@ -116,16 +138,28 @@ const Sharect = (function(){
...
+        document.body.appendChild(div)
+      }
     }
 
     function attachEvents() {
-      function hasSelection() {
-        return !!window.getSelection().toString()
-      }
 
       function hasTooltipDrawn() {
```
Comment (estevanmaito) [inline src/sharect.js:150]: Avoid side effects.
This function should return just `true` or `false`. If DOM manipulation or any other change is needed, it should be done outside of this function.
The `mouseup` event already tests for tooltip drawn and if it has selection. Maybe you could add another test there to look at its `offsetParent`.

**Review by estevanmaito — CHANGES_REQUESTED** [review]
Thanks for your time!
I really appreciate your changes. I've just added some comments along the modifications in the src code. Please, take a look.

File: src/sharect.js (hunk 6 above)
Comment (Laurian) [reply src/sharect.js:57]: true, slipped in as I needed it, will revert it in my branch

**Comment by Laurian** [comment]
OK, I will do the changes, I might also fix the calculation of the position if there are nested positioned ancestors, the current fix deals with one only.

**Comment by estevanmaito** [comment]
Even though mispositioned tooltip is an issue inside nested elements, it's an edge case that has never become an issue, even more than a year after the library launch.
Besides that, the changes asked weren't made, and it still doesn't solve the problem of nested elements.

**Pull request closed without merging**